website/blog/2021-08-18-improving-marker-mechanism.md

```diff
@@ -0,0 +1,65 @@
+
+Next, we dive into the existing marker mechanism, explain its performance problem, and demonstrate the new timeline-server-based marker mechanism to address the problem.
+
+## Existing direct marker file mechanism and its limitations
+
+The **existing marker mechanism** simply creates a new marker file corresponding to each data file, with the marker filename as described above.  Each marker file is written to the file system in the same directory hierarchy, i.e., commit instant and partition path, under a temporary folder `.hoodie/.temp` under the base path of the Hudi table.  For example, the figure below shows one example of the marker files created and the corresponding data files when writing data to the Hudi table.  When getting or deleting all the marker file paths, the mechanism first lists all the paths under the temporary folder, `.hoodie/.temp/<commit_instant>`, and then does the operation.
+
+![An example of marker and data files in direct marker file mechanism](/assets/images/blog/marker-mechanism/direct-marker-file-mechanism.png)
+
+As the number of data files to write increases, so does the number of marker files to create.  This can create performance bottlenecks for cloud storage such as AWS S3.  In AWS S3, each file create and delete call triggers an HTTP request and there is [rate-limiting](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html) on how many requests can be processed per second per prefix in a bucket.  When the number of data files to write concurrently and the number of marker files is huge, the marker file operations become the performance bottleneck.  In one case, the marker file deletion takes an hour to finish due to S3 rate-limiting for a bulk insert operation running for a few hours.  Such behavior degrades the performance of the write.
+
+## Timeline-server-based marker mechanism improving write performance
+
+To address the performance bottleneck due to rate-limiting of AWS S3 explained above, we introduce a **new marker mechanism leveraging the timeline server**, which optimizes the marker-related latency for file systems with non-trivial file I/O latency.  The **timeline server** in Hudi serves as a centralized place for providing the file system and timeline views. As shown below, the new timeline-server-based marker mechanism delegates the marker creation and other marker-related operations from individual executors to the timeline server for centralized processing.  The timeline server maintains the created markers in memory for corresponding marker requests.  The timeline server achieves consistency by periodically flushing the in-memory markers to a limited number of underlying files in the file system.  In such a way, the number of actual file operations and latency related to markers can be significantly reduced even with a huge number of data files, thus improving the performance of the writes.
```

Comment:
"The timeline server maintains the created markers in memory for corresponding marker requests.  The timeline server achieves consistency by periodically flushing the in-memory markers to a limited number of underlying files in the file system." -> 
"The timeline server batches the marker creation requests and writes to a bounded set of marker files at regular intervals."